### PR TITLE
part of #1458, enable secured HBase

### PR DESCRIPTION
**Pull request by huafengw**

### CODE REVIEW

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+object HBaseSecurityUtil {
+  val LOG: Logger = LogUtil.getLogger(getClass)
+
+  def getConnection(userConfig: UserConfig, configuration: Configuration): Connection = {
+    if(UserGroupInformation.isSecurityEnabled) {
+      val principal = userConfig.getString(Constants.GEARPUMP_KERBEROS_PRINCIPAL)
+      if(principal.isEmpty) {
+        LOG.error(s"User's kerberos principal is not provided")
+      }
+      val keytabContent = userConfig.getBytes(Constants.GEARPUMP_KEYTAB_FILE)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: why is this called keytabContent while the config key is `GEARPUMP_KEYTAB_FILE`

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      if(keytabContent.isEmpty) {
+        LOG.error(s"Keytab file is not provided")
+      }
+      val keytabFile = File.createTempFile("login", ".keytab")
+      FileUtils.writeByteArrayToFile(keytabFile, keytabContent.get)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: what if keytabContent is empty ?

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      keytabFile.setExecutable(false)
+      keytabFile.setWritable(false)
+      keytabFile.setReadable(true, true)
+
+      UserGroupInformation.setConfiguration(configuration)
+      UserGroupInformation.loginUserFromKeytab(principal.get, keytabFile.getAbsolutePath)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: and what if principal is empty ?

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      keytabFile.delete()
+    }
+    ConnectionFactory.createConnection(configuration)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: new connection for every call or reuse ?

File: core/src/main/scala/io/gearpump/util/Constants.scala [hunk 5]
```diff
@@ -123,4 +123,8 @@ object Constants {
     classOf[HashPartitioner],
     classOf[ShuffleGroupingPartitioner],
     classOf[ShufflePartitioner])
+
+  //Security related
```
Comment (manuzhang) [inline core/src/main/scala/io/gearpump/util/Constants.scala]: related to all components or just HBase ?

Comment (huafengw) [inline core/src/main/scala/io/gearpump/util/Constants.scala]: I think one pair of authenticated principal and keytab is enough for all components.

Comment (manuzhang) [inline core/src/main/scala/io/gearpump/util/Constants.scala]: Could we have more information here like what kind of security and how the duo are supposed to be used ?

Comment (huafengw) [inline core/src/main/scala/io/gearpump/util/Constants.scala]: Add a README

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        throw new Exception(errorMsg)
+      }
+      val keytabFile = File.createTempFile("login", ".keytab")
+      FileUtils.writeByteArrayToFile(keytabFile, keytabContent.get)
+      keytabFile.setExecutable(false)
+      keytabFile.setWritable(false)
+      keytabFile.setReadable(true, true)
+
+      UserGroupInformation.setConfiguration(configuration)
+      UserGroupInformation.loginUserFromKeytab(principal.get, keytabFile.getAbsolutePath)
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: Is `IOException` handled here ?

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  def getConnection(userConfig: UserConfig, configuration: Configuration): Connection = {
+    if(UserGroupInformation.isSecurityEnabled) {
+      val principal = userConfig.getString(Constants.GEARPUMP_KERBEROS_PRINCIPAL)
+      val keytabContent = userConfig.getBytes(Constants.GEARPUMP_KEYTAB_FILE)
+      if(principal.isEmpty || keytabContent.isEmpty) {
+        val errorMsg = s"HBase is security enabled, user should provide kerberos principal in " +
+          s"${Constants.GEARPUMP_KERBEROS_PRINCIPAL} and keytab file in ${Constants.GEARPUMP_KEYTAB_FILE}"
+        throw new Exception(errorMsg)
+      }
+      val keytabFile = File.createTempFile("login", ".keytab")
```
Comment (manuzhang) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: Is `IOException` handled for all the file interactions ?

Comment (huafengw) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: Is it better to just throw the exceptions?

File: external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+  def getConnection(userConfig: UserConfig, configuration: Configuration): Connection = {
+    if(UserGroupInformation.isSecurityEnabled) {
+      val principal = userConfig.getString(Constants.GEARPUMP_KERBEROS_PRINCIPAL)
+      val keytabContent = userConfig.getBytes(Constants.GEARPUMP_KEYTAB_FILE)
+      if(principal.isEmpty || keytabContent.isEmpty) {
+        val errorMsg = s"HBase is security enabled, user should provide kerberos principal in " +
+          s"${Constants.GEARPUMP_KERBEROS_PRINCIPAL} and keytab file in ${Constants.GEARPUMP_KEYTAB_FILE}"
+        throw new Exception(errorMsg)
+      }
```
Comment (clockfly) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: It is not safe to store keytab file under temporary folder.

Comment (huafengw) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: store under ```System.getProperty("user.home")```?

Comment (clockfly) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]: It is better, but not solving the problem.

With no access control, it is still possble for other user to access it.

Comment (huafengw) [inline external/hbase/src/main/scala/io/gearpump/external/hbase/HBaseSecurityUtil.scala]:
```
keytabFile.setExecutable(false)
keytabFile.setWritable(false)
keytabFile.setReadable(true, true)
```
You mean this kind of access control?

**Comment by clockfly** [comment]
Will this code also work for HBase cluster without security enabled?

**Comment by huafengw** [comment]
Yes, already tested.

**Comment by clockfly** [comment]
+1

**Pull request merged**